### PR TITLE
Add possibility to manually unsubscribe from useLazyQuery #2055

### PR DESCRIPTION
**Pull request by damianborowy**
Related to issue [#2055](https://github.com/reduxjs/redux-toolkit/issues/2055)

### CODE REVIEW

File: packages/toolkit/src/query/react/buildHooks.ts [hunk 1]
```diff
@@ -835,16 +836,16 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
...
           ...options,
           skip: arg === UNINITIALIZED_VALUE,
         })
 
         const info = useMemo(() => ({ lastArg: arg }), [arg])
         return useMemo(
-          () => [trigger, queryStateResults, info],
-          [trigger, queryStateResults, info]
+          () => [trigger, queryStateResults, info, unsubscribe],
+          [trigger, queryStateResults, info, unsubscribe]
```
Comment (phryneas) [inline packages/toolkit/src/query/react/buildHooks.ts:848]: I would suggest to make `unsubscribe` part of the `info` object and maybe give it a better name. Open to suggestions.

Comment (damianborowy) [reply packages/toolkit/src/query/react/buildHooks.ts]: That's some good advice, I missed the fact that `info` was already an object for some extra things so I can touch it without breaking anything 😄 not sure if `extras` is the best possible name though, waiting for further feedback!

**Comment by phryneas** [comment]
We just did a grooming of PRs and discussed this: would you be okay to to a PR for a `reset` method instead that unsubscribed *and* set the query hook back to `isUninitialized`?

It feels inconsistent to us that after unsubscribing, the query state will still stay in the component for another 60 seconds - or even forever, if another component is still using it.

Or do you have a good use case that would only make sense with `unsubscribe`, but not with `reset`?

**Comment by henkkasoft** [comment]
> `reset`

This is exactly what I am looking for I think.

I am using useLazyQuery with query params given by user. 
The API response is HTTP ERROR 422 when the user input is incorrectly formed.
UI will show modal of the error and then I want to recover without unmounting the component. So I would want get rid of the isError state and the user could modify the input and trigger the query again.

Is this possible somehow already? 
And if not, did I understood correctly that this PR (when unsubscribe  is extended as reset) will enable this kind of usage?

**Comment by henkkasoft** [comment]
> We just did a grooming of PRs and discussed this: would you be okay to to a PR for a `reset` method instead that unsubscribed _and_ set the query hook back to `isUninitialized`?
> 
> It feels inconsistent to us that after unsubscribing, the query state will still stay in the component for another 60 seconds - or even forever, if another component is still using it.
> 
> Or do you have a good use case that would only make sense with `unsubscribe`, but not with `reset`?

This is exactly what I am looking for I think.

I am using useLazyQuery with query params given by user.
The API response is HTTP ERROR 422 when the user input is incorrectly formed.
UI will show modal of the error and then I want to recover without unmounting the component. So I would want get rid of the isError state and the user could modify the input and trigger the query again.

Is this possible somehow already?
And if not, did I understood correctly that this PR (when unsubscribe is extended as reset) will enable this kind of usage?

**Comment by phryneas** [comment]
@henkkasoft it will not remove the value, just unsubscribe from it. It will be removed after 60 second if no other component uses it.

Maybe you are giving `isError` too much importance here?  
As it sounds "there is an error from the server" will 100% be true until the data is sent to the server without a 422 - so maybe you just need an additional local state to control that message?

**Comment by stchristian** [comment]
Hi, I see that there was no activity on this PR for a while now. Is this still pending? I would be happy to take over this PR and implement the above mentioned `reset` method
@damianborowy @henkkasoft

**Comment by markerikson** [comment]
@stchristian yeah, no one has looked at this in a year. Might be best to re-create it and file a new PR with the updates

**Comment by stchristian** [comment]
@markerikson Before starting this feature, I would suggest adding a `skip` parameter instead of `reset` method, similar to `useQuery()` that would work the same way setting the hook to `uninitialized`. This way the query stays consistent with the existing interface.

**Comment by henkkasoft** [comment]
I was able to prevent invalidation of the query in isError state by adding error check to providesTags function:
`return error ? [] : ['MyTag']`
Now error state queries are not invalidated when 'MyTag' is invalidated in some mutation's invalidatesTags function.

Not sure if this is still the best solution...

**Comment by markerikson** [comment]
Replaced by #4689

**Pull request closed without merging**